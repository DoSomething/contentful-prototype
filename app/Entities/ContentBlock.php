<?php

namespace App\Entities;

use JsonSerializable;

class ContentBlock extends Entity implements JsonSerializable
{
    /**
     * Convert the object into something JSON serializable.
     *
     * @return array
     */
    public function jsonSerialize()
    {
        return [
            'id' => $this->entry->getId(),
            'type' => $this->getContentType(),
            'fields' => [
                'superTitle' => $this->superTitle,
                'title' => $this->title,
                'subTitle' => $this->subTitle,
                'content' => $this->content,
                'image' => get_image_url($this->image, 'square'),
                'imageAlignment' => $this->imageAlignment ? strtolower($this->imageAlignment) : null,
                'additionalContent' => $this->additionalContent,
            ],
        ];
    }
}
