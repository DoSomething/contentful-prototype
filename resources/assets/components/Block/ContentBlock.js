import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { PuckWaypoint } from '@dosomething/puck-client';

import SectionHeader from '../SectionHeader';
import Markdown from '../Markdown';
import { Flex, FlexCell } from '../Flex';

import './contentBlock.scss';

// TODO: Replace alt with better description.
const renderPhoto = (photo, index) => (
  <div className="content-block__photo" key={index}>
    <img alt="Action step" src={photo} />
  </div>
);

const ContentBlock = props => {
  const {
    title,
    stepIndex,
    content,
    photos,
    photoWidth,
    shouldTruncate,
    hideStepNumber,
    preTitle,
  } = props;

  const photoComponent =
    photos && photos.length ? (
      <FlexCell width={photoWidth}>
        <div className={`content-block__photos -${photoWidth}`}>
          {photos ? photos.map(renderPhoto) : null}
        </div>
      </FlexCell>
    ) : null;

  return (
    <div
      className={classnames('content-block', { '-truncate': shouldTruncate })}
    >
      <Flex>
        <SectionHeader
          preTitle={preTitle}
          title={title}
          step={stepIndex}
          hideStepNumber={hideStepNumber}
        />
        {content ? (
          <FlexCell width="two-thirds">
            <Markdown>{content}</Markdown>
          </FlexCell>
        ) : null}
        {photoComponent}
      </Flex>
      <PuckWaypoint name="content-block__bottom" waypointData={{ title }} />
    </div>
  );
};

ContentBlock.propTypes = {
  preTitle: PropTypes.string,
  title: PropTypes.string.isRequired,
  stepIndex: PropTypes.number.isRequired,
  content: PropTypes.string,
  photos: PropTypes.arrayOf(PropTypes.string),
  photoWidth: PropTypes.string.isRequired,
  shouldTruncate: PropTypes.bool,
  hideStepNumber: PropTypes.bool,
};

ContentBlock.defaultProps = {
  preTitle: null,
  content: null,
  photos: [],
  shouldTruncate: false,
  hideStepNumber: false,
};

export default ContentBlock;
