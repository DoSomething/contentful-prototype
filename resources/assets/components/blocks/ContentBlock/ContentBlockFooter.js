import React from 'react';
import PropTypes from 'prop-types';

import ErrorBlock from '../ErrorBlock/ErrorBlock';
import VoterWidgetBlock from '../VoterWidgetBlock/VoterWidgetBlock';
import GetOutTheVoteBlock from '../GetOutTheVoteBlock/GetOutTheVoteBlock';

const ContentBlockFooter = ({ type }) => {
  switch (type) {
    case 'VoterWidgetBlock':
      return <VoterWidgetBlock />;
    case 'GetOutTheVoteBlock':
      return <GetOutTheVoteBlock />;

    default:
      return <ErrorBlock error={`ContentBlock is unable to render ${type}.`} />;
  }
};

ContentBlockFooter.propTypes = {
  type: PropTypes.string,
};

ContentBlockFooter.defaultProps = {
  type: null,
};

export default ContentBlockFooter;
