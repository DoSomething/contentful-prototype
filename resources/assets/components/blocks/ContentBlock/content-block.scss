@import '../../../scss/next-toolbox.scss';

.content-block {
  width: 100%;
  margin-bottom: 36px;

  h1,
  h2,
  h3,
  p {
    color: $black;
    font-family: $primary-font-family;
  }

  .two-thirds {
    @include media($medium) {
      width: 66.6666666666666%;
    }
  }
}
