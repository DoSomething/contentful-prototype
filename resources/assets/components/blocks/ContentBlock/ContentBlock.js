import React from 'react';
import PropTypes from 'prop-types';

import { Figure } from '../../Figure';
import { Flex, FlexCell } from '../../Flex';
import SectionHeader from '../../SectionHeader';
import Markdown from '../../utilities/Markdown/Markdown';

import './content-block.scss';

const ContentBlock = props => {
  const { content, image, imageAlignment, superTitle, title } = props;

  const contentNode = content ? <Markdown>{content}</Markdown> : null;

  return (
    <div className="content-block">
      {title ? (
        <div className="margin-horizontal-md">
          <SectionHeader preTitle={superTitle} title={title} hideStepNumber />
        </div>
      ) : null}

      <div className="margin-horizontal-md">
        {image ? (
          <Figure
            image={image}
            alt="content-block"
            alignment={`${imageAlignment}-collapse`}
            size="one-third"
          >
            {contentNode}
          </Figure>
        ) : (
          <Flex>
            <FlexCell width="two-thirds">{contentNode}</FlexCell>
          </Flex>
        )}
      </div>
    </div>
  );
};

ContentBlock.propTypes = {
  content: PropTypes.string.isRequired,
  image: PropTypes.string,
  imageAlignment: PropTypes.string,
  superTitle: PropTypes.string,
  title: PropTypes.string,
};

ContentBlock.defaultProps = {
  image: null,
  imageAlignment: 'right',
  superTitle: null,
  title: null,
};

export default ContentBlock;
