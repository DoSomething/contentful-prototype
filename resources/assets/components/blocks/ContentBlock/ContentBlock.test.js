import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';

import ContentBlock from './ContentBlock';

const emptyImage = { url: null, description: null };

const props = {
  superTitle: 'Test Super Title',
  title: 'Test Title',
  content: 'Test Content',
  image: {
    url: 'http://image.com',
    description: 'cool image of http://image.com',
  },
  imageAlignment: 'RIGHT',
};

describe('ContentBlock component', () => {
  test('is rendered with the proper child components when image is set', () => {
    const wrapper = shallow(<ContentBlock {...props} />);

    expect(wrapper.find('SectionHeader').length).toEqual(1);
    expect(wrapper.find('LazyImage').length).toEqual(1);
    expect(wrapper.find('TextContent').length).toEqual(1);

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  test("does not include SectionHeader when there's no title", () => {
    const wrapper = shallow(<ContentBlock {...props} title={undefined} />);

    expect(wrapper.find('SectionHeader').length).toEqual(0);

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  test('is rendered with the proper child components when image is not set', () => {
    const wrapper = shallow(<ContentBlock {...props} image={emptyImage} />);

    expect(wrapper.find('SectionHeader').length).toEqual(1);
    expect(wrapper.find('LazyImage').length).toEqual(0);
    expect(wrapper.find('TextContent').length).toEqual(1);
  });

  test('it works beautifully with content and an empty image prop', () => {
    const wrapper = shallow(
      <ContentBlock content="hi there" image={emptyImage} />,
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe('Content column span', () => {
    test('it displays across the full row if there is no image & the fullWidthContent prop is toggled on', () => {
      render(<ContentBlock {...props} image={emptyImage} fullWidthContent />);

      expect(screen.getByTestId('content-block-content').className).toContain(
        'col-span-3',
      );
    });

    test('it displays across two-thirds of the row if an image is not provided but the fullWidthContent is not toggled on', () => {
      render(<ContentBlock {...props} image={emptyImage} />);

      const contentBlockContent = screen.getByTestId('content-block-content');
      expect(contentBlockContent.className).toContain('col-span-2');
      expect(contentBlockContent.className).not.toContain('col-span-3');
    });

    test('it displays across two-thirds of the row when an image *is* provided, even if the fullWidthContent is toggled on', () => {
      render(<ContentBlock {...props} fullWidthContent />);

      const contentBlockContent = screen.getByTestId('content-block-content');
      expect(contentBlockContent.className).toContain('col-span-2');
      expect(contentBlockContent.className).not.toContain('col-span-3');
    });
  });
});
