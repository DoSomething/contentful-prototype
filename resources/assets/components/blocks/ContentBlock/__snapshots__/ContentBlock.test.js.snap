// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentBlock component does not include SectionHeader when there's no title 1`] = `
<div
  className="content-block"
>
  <div
    className="margin-horizontal-md"
  >
    <Figure
      alignment="right-collapse"
      alt="content-block"
      image="image.com?w=600&h=600&fit=fill"
      imageClassName={null}
      size="one-third"
    >
      <Markdown
        className={null}
      >
        Test Content
      </Markdown>
    </Figure>
  </div>
</div>
`;

exports[`ContentBlock component is rendered with the proper child components when image is set 1`] = `
<div
  className="content-block"
>
  <div
    className="margin-horizontal-md"
  >
    <SectionHeader
      preTitle="Test Super Title"
      title="Test Title"
    />
  </div>
  <div
    className="margin-horizontal-md"
  >
    <Figure
      alignment="right-collapse"
      alt="content-block"
      image="image.com?w=600&h=600&fit=fill"
      imageClassName={null}
      size="one-third"
    >
      <Markdown
        className={null}
      >
        Test Content
      </Markdown>
    </Figure>
  </div>
</div>
`;

exports[`ContentBlock component it works beautifully with a mere content prop 1`] = `
<div
  className="content-block"
>
  <div
    className="margin-horizontal-md"
  >
    <Markdown
      className={null}
    >
      hi there
    </Markdown>
  </div>
</div>
`;
